src/pieces.py

```diff
@@ -14,6 +26,7 @@ def valid_steps(self):
         Raises:
             NotImplementedError: Needs implementation.
         """
+
```

Comment:
All these new whitespaces are violating [D202](https://www.pydocstyle.org/en/2.1.1/error_codes.html)

Reply:
You mean one blank line after a docstring? What about [D204](https://www.pydocstyle.org/en/6.1.1/error_codes.html)? Is there a reason class docstrings have an empty line, while function docstrings do not? I remind that multiline functions are allowed single blank lines in their body, therefore it make is very weird to stick up their first line, whatever it is, with the function's docstring? Isn't iniformity an issue here? I am surprised with the differentiation as in D202 and D204. Especially if the function docstring is multiline too (always is basically).

Reply:
Yes. Please no newlines after a method's docstring. Also, please add a newline after a class's docstring. Let's follow both D202 and D204.

Reply:
We don't want to be flooded by newlines and methods are a commodity. Class docstrings are rare and are usually followed by a method. We separate methods in a class by a newline anyway.

Reply:
I still think it is very weird. I have no issue complying with PEP8, however you do see the issue of a docstring full of blank lines and a code body full of blank lines to be zipped tight on their last and first line respectively. Please go ahead and fix the docstrings, this is one of those single commit PRs (clear cut case). I still disagree, but PEP8 is what it is.

Reply:
> We don't want to be flooded by newlines and methods are a commodity. Class docstrings are rare and are usually followed by a method. We separate methods in a class by a newline anyway.

So basically the D204 new line is simply the class's first method overhead blank line, I see. Very well. Please feel free to make this determinist correction yourself if you may, thank you.

Reply:
I think there is good separation because in most cases there is an entire line with just `"""` plus syntax highlighting distinguishes them considerably.

btw that is actually [PEP-257](https://peps.python.org/pep-0257/).

Reply:
All right, we are in agreement.

Reply:
I would prefer if you did it, since you inserted the newlines. Just enable `pydocstyle` in your IDE and you will find them all easily.

Reply:
Surprisingly, they were not marked, and as far as I know my pydocstlye has no exceptions, though now I am not so sure. This needs checking first, before I submit a correcting commit.